### DISCUSSION TITLE
Trying to write blended pixels to the screen, getting no transparency / black clear.

### DISCUSSION THREAD

**Discussion by Zageron** [parasyte/pixels#293] [Q&A] [answered]
Disclaimer: I am not trying to render the window as transparent, though I may end up doing that, right now I'm just trying to blit a texture with transparency information onto the screen.

Screen is standard `screen: &mut [u8]` from examples.

![image](https://user-images.githubusercontent.com/1892473/181859430-4593055c-2313-4d4e-9720-129f2f056471.png)
![image](https://user-images.githubusercontent.com/1892473/181859448-eb7d6cba-cd2c-46ae-9e4d-d7fff545d7d5.png)
![image](https://user-images.githubusercontent.com/1892473/181859471-b4b8f0d7-c61e-4410-a8d5-b38894e99d5c.png)

Wondering if the format is conflicting, the screen shows "cleared" as `\xff\xff\xff,`, so I am a little confused as to the format it is expecting.

I also tried the following, with no luck.
```rs
let pixels = if let Ok(pixels) = PixelsBuilder::new(WIDTH, HEIGHT, surface_texture)
        .surface_texture_format(TextureFormat::Bgra8UnormSrgb)
        .build_async()
        .await
```

This is the result. (I've tried different alpha blending modes, with no avail so not sure.) (The colors are correct, just no transparency / blending.)
![image](https://user-images.githubusercontent.com/1892473/181859994-38e4f9a7-b02d-4869-8edf-0ab3ce5cbacf.png)

These are my image conversion options: (I am currently using to_rgba8.)
![image](https://user-images.githubusercontent.com/1892473/181860416-ad3af435-5189-418e-b2e3-76bf6a9f4fa0.png)

**Comment by parasyte** [comment, marked as answer]
If I understand correctly what you are doing, there is no blending in the latest released version so you are going to see the exact color values you put into the buffer. The alpha channel is completely ignored.

The unreleased main branch has configurable `BlendState`, and defaults to alpha blending, which is probably closer to what you expect.

I don’t know what you mean by “alpha blending modes”, that seems to imply you were trying to change the `BlendState` on the unreleased main branch, and it was having no affect. Is this the case?

**Reply by Zageron** [reply]
I am currently working off of `pixels = { git = "https://github.com/JMS55/pixels.git", branch = "main" }`, which has the options you mentioned.
I meant these settings by "alpha blending modes": `.blend_state(BlendState::ALPHA_BLENDING)`.

**Reply by Zageron** [reply]
Code in question is here:
https://github.com/flashflashrevolution/rrr/blob/blending/crates/head/src/sprites.rs
https://github.com/flashflashrevolution/rrr/blob/blending/crates/head/src/main.rs#L376

**Reply by Zageron** [reply]
I will try a downgrade to older wgpu and use the master branch of pixel to see if that resolves the issue.

**Reply by parasyte** [reply]
That sounds like a bug, if `BlendState::ALPHA_BLENDING` does not blend! Unless something unusual is happening on your end and the pixel buffer does not receive the alpha values that you expect.

**Reply by Zageron** [reply]
Tried the downgrade, no luck.
The pixel buffer is receiving the exact colours from the image.

**Reply by Zageron** [reply]
The RGB channels are working correctly, the alpha doesn't seem to be.

**Reply by Zageron** [reply]
Source png, with alpha channel.
![image](https://user-images.githubusercontent.com/1892473/181875312-bb5c959f-4ff5-44b4-bb8d-9f01fb1ac2a9.png)

Library used to import is rust image. Saving back out after modifications retains alpha channel.

Here's my logs in case that gives any clarity.
```
2022-07-30T05:01:17.016Z INFO  [wgpu_core::device] Created texture Valid((0, 1, Vulkan)) with TextureDescriptor { label: Some("pixels_source_texture"), size: Extent3d { width: 512, height: 768, depth_or_array_layers: 1 }, mip_level_count: 1, sample_count: 1, dimension: D2, format: Rgba8UnormSrgb, usage: COPY_DST | TEXTURE_BINDING }
2022-07-30T05:01:17.024Z INFO  [wgpu_core::device] Created buffer Valid((0, 1, Vulkan)) with BufferDescriptor { label: Some("pixels_scaling_renderer_vertex_buffer"), size: 24, usage: VERTEX, mapped_at_creation: true }
2022-07-30T05:01:17.026Z INFO  [wgpu_core::device] Created buffer Valid((1, 1, Vulkan)) with BufferDescriptor { label: Some("pixels_scaling_renderer_matrix_uniform_buffer"), size: 64, usage: COPY_DST | UNIFORM, mapped_at_creation: true }
2022-07-30T05:01:17.067Z INFO  [wgpu_core::device] configuring surface with SurfaceConfiguration { usage: RENDER_ATTACHMENT, format: Bgra8UnormSrgb, width: 512, height: 768, present_mode: Fifo }
2022-07-30T05:01:17.136Z INFO  [wgpu_core::device] configuring surface with SurfaceConfiguration { usage: RENDER_ATTACHMENT, format: Bgra8UnormSrgb, width: 1400, height: 840, present_mode: Fifo }
2022-07-30T05:01:17.142Z WARN  [wgpu_core::device] Requested size 1400x840 is outside of the supported range: Extent3d { width: 512, height: 768, depth_or_array_layers: 1 }..=Extent3d { width: 512, height: 768, depth_or_array_layers: 1 }
2022-07-30T05:01:17.144Z INFO  [wgpu_hal::vulkan::instance] GENERAL [NVIDIA (0x4)]
        Requested image extent (1400x840) does not match surface (512x768), marking swapchain out of date
2022-07-30T05:01:17.144Z INFO  [wgpu_hal::vulkan::instance]     objects: (type: SWAPCHAIN_KHR, hndl: 0x20ff12423e0, name: ?)
2022-07-30T05:01:17.146Z INFO  [wgpu_core::device] configuring surface with SurfaceConfiguration { usage: RENDER_ATTACHMENT, format: Bgra8UnormSrgb, width: 512, height: 768, present_mode: Fifo }
2022-07-30T05:01:17.152Z INFO  [wgpu_core::device] configuring surface with SurfaceConfiguration { usage: RENDER_ATTACHMENT, format: Bgra8UnormSrgb, width: 512, height: 768, present_mode: Fifo }
```

**Reply by Zageron** [reply]
Please let me know if there is anything I can do to help you understand what might be wrong.
Or, if you have any advice on what I can do to better discover how I might fix this.

**Reply by parasyte** [reply]
Are you doing your own blending? This might seem like a silly question, but `pixels` is not going to do _any_ blending when you write color data to the pixel buffer. You're 100% in control of what you put into it.

The `BlendState` works with texture compositing on the surface. That is, the pixel buffer is a texture containing an alpha channel, and the surface is a texture which does not (with `pixels` the surface texture only provides the background color). Assuming your pixel buffer is mostly transparent like your source png, you will be able to set the clear color to whatever you want and `wgpu` will blend your pixel buffer with it, producing the kind of effect that I believe you are after.

That said, your example has a lot of code which is irrelevant to the investigation. The common practice is to share a minimal reproduction case to demonstrate that there is a bug. Since I am not familiar with your code base and there is far too much for me to pick through, I will instead start from scratch and build a small demo that shows blending working as intended.

Even this was a bit painful. I wanted the demo to show there is "nothing up my sleeve" so the background color uses a simple linear animation from black to white and back. I don't have a copy of your source image, but I don't think that is relevant to the issue, either. I just created an image with some colored text on a transparent background.

The demo code is here: https://github.com/parasyte/alpha-test

The main point of interest is that whatever you put into the pixel buffer with `alpha < 1` will let some of the background show through. The background can only be a solid color (called the clear color). If you want to do any more advanced blending like compositing two images together, that's up to you in your blit function. Based on your comments, I don't think you are that far along, yet.

Given that my demo works (for me, at least!) it doesn't seem there is a bug on our end.

**Reply by parasyte** [reply]
I suspect what is happening in your code is precisely what I alluded to. You are not doing any blending in your blit, but you expect `pixels` to do that.

1. The pixel buffer is cleared with [white and 17% alpha](https://github.com/flashflashrevolution/rrr/blob/262c014ec00e903e7bcf4df0d09a9cbf25331478/crates/head/src/main.rs#L169).
2. [Arrows are blitted to the buffer](https://github.com/flashflashrevolution/rrr/blob/262c014ec00e903e7bcf4df0d09a9cbf25331478/crates/head/src/main.rs#L196) without any compositing at all, wiping out the existing translucent white.
    - The [`blit()`](https://github.com/flashflashrevolution/rrr/blob/262c014ec00e903e7bcf4df0d09a9cbf25331478/crates/head/src/sprites.rs#L13-L73) function is responsible for doing the compositing.

If you take the pixel buffer contents after `game.draw()` returns and write it to a png file, you will see exactly what has gone wrong. The image will look exactly like what you see in the window! Except the black area around the arrows will be fully transparent (`alpha = 0`) and everything around it will be translucent white. What you are seeing in the window is a blend of this "broken" image with the clear color (which is black).

You can fix it by replacing your `clear()` function with one that clears the buffer to transparent black (like what I have in the demo) and set the clear color to the final _composited_ background color that you want. Assuming that is a mid gray, as in the translucent white blended with black, you would do this:

```rust
let intensity = (44.0 / 255.0).powf(2.2);

PixelsBuilder::new(WIDTH, HEIGHT, surface_texture)
    .clear_color(pixels::wgpu::Color {
        r: intensity,
        g: intensity,
        b: intensity,
        a: 1.0,
    })
    .build_async()
    .await
```

That should produce more-or-less exactly what you want. The background will be approximately mid-gray as shown in your first screenshot.

**Reply by Zageron** [reply]
Thank you very much for all of the details, just based on your explanation I am 100% certain that my issue is exactly as you described. The fact that it didn't automatically blend with the manual clear color went straight over my head.

Thank you very much for the example, I will read through it shortly and hopefully I'm smart enough to solve my problem.

Thanks again for spending all that time writing the example, by the way, that was very generous of you.

**Reply by parasyte** [reply]
My pleasure! I have an obligation to ensure that `pixels` functions as intended. If it can help a user, that's even better. :) No wasted effort on my part.

If there is anything else I can do, I'll gladly try my best.
